044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,295 @@
+Additionally, it ties the plug-in directly to storage usage, there are other factors which users may wish to respond to.
+
+As clusters scale up the likelihood of even topic distribution drops. When topics are not evenly distributed it
+is possible for replication from broker `1` to broker `2` will cause broker `2` to consume all available storage
+without triggering throttling of clients, as broker `1` disk usage remains acceptable.
+
+Addressing the effects of uneven topic distribution sounds like it should come
+under [KIP-73](https://cwiki.apache.org/confluence/display/KAFKA/KIP-73+Replication+Quotas). Unfortunately replication
+quotas are designed to manage the additional network load of migrating replicas between brokers, which does not address
+client publication leading to out of disk conditions through replication. As replication throttling is configured in terms
+of `TopicPartitions` the quota plug-in would need a mechanism to translate a logDir into a set of TopicPartitions.
+Assuming it could select generate the appropriate throttle configuration this could lead to unpredictable latency spikes
+for producers configured with `acks >= 1` as the replication from partition leader to follower maybe delayed due to the 
+throttle.
+
+Currently, the kafka-quotas-plugin considers the total quantity of storage and how much of that is used (
+see [issue#2](https://github.com/strimzi/kafka-quotas-plugin/issues/2)) when considering whether to apply throttling to
+clients. This is problematic with respect to handling disk failure for Just a Bunch Of Disks (JBOD) deployments [(KIP-112)](https://cwiki.apache.org/confluence/display/KAFKA/KIP-112%3A+Handle+disk+failure+for+JBOD) as the
+broker will take partitions offline when the volume they are stored on runs out of space. Which in the case of unbalanced usage
+between volumes can lead to a volume running out of storage without throttling being applied.
+
+Throttling down message production on all broker nodes will protect the cluster from
+running out of disk due to replication.
+
+## Motivation
+
+Users need better protection against running out of disk on any of their nodes as it can degrade the service
+in unpredictable ways (corrupted segment logs) which could impede the interventions required to recover.
+
+## Proposal
+
+1. Extend the kafka-quotas-plugin so that we can choose to observe the disk usage 
+of all brokers in the cluster.
+2. [Deprecate](#compatibility) the existing local disk observations.
+3. Add new limit types, so that it is explicit what is being limited and better support heterogeneous disks
+   1. throttle if used bytes exceeds threshold on any volume
+   2. throttle if available bytes less than threshold on any volume
+   3. throttle if available ratio less than threshold on any volume
+4. Introduce extension points in the quotas-plugin to support pluggable sources of quotas and throttle factors
```

Comment:
Do we really want to have plugin to a plugin? Would designing the API to support subclassing be a better alternative?

Reply:
I've tried to update the language used to avoid the idea of additional plug-ins.